### PR TITLE
Sliding/v1

### PR DESCRIPTION
**Pull request by cryogenian**
[ch13298]

I'm a bit unhappy with `build` using `Stream`, and nothing producing this `Stream` with more than one element (in practice this should be the case, incremental loads shouldn't produce more than one stage file). But this pr definitely imporves performance back to pre-temp-table approach for create.

### CODE REVIEW

**Comment by rintcius** [comment]
FYI Will review today

File: core/src/main/scala/quasar/destination/snowflake/TempTable.scala [hunk 1]
```diff
@@ -0,0 +1,255 @@
...
+        case _ => MonadResourceErr[F].raiseError(ResourceError.notAResource(args.path))
+      }
+      createColumnFragment <- args.columns.traverse(mkColumn(hygienicIdent, _)).fold(
+        errs => Sync[F].raiseError(ColumnTypesNotSupported(errs)),
+        xs => Fragments.parentheses(xs.intercalate(fr",")).pure[F])
+      _ <- checkWriteMode(tbl)
+      refMode <- Ref.of[F, QWriteMode](args.writeMode)
+    } yield new Builder[F] {
+
+      val tmpName = s"precog_tmp_$tbl"
```
Comment (rintcius) [inline core/src/main/scala/quasar/destination/snowflake/TempTable.scala:94]: Worth postfixing with a random uuid or something or is this sufficient?

File: core/src/main/scala/quasar/destination/snowflake/SnowflakeDestination.scala [hunk 2]
```diff
@@ -19,45 +19,47 @@ package quasar.destination.snowflake
...
     SnowflakeDestinationModule.destinationType
 
-  def flowResource(args: FlowArgs[ColumnType.Scalar]): Resource[F, Flow[Byte]] =
-    TempTableFlow(xa, logger, writeMode, schema, hygienicIdent, args) map { (flow: Flow[Byte]) =>
-      flow.mapK(Retry[F](maxRetries, retryTimeout))
-    }
+  def blocker: Blocker =
+    Blocker.liftExecutionContext(
+      ExecutionContext.fromExecutor(
+        Executors.newCachedThreadPool(NamedDaemonThreadFactory("snowflake-destination"))))
```
Comment (rintcius) [inline core/src/main/scala/quasar/destination/snowflake/SnowflakeDestination.scala:57]: Possible to create this once in the module, as a `val` and passing it as an argument to `SnowflakeDestination` (similar as `logger` basically)?

File: core/src/main/scala/quasar/destination/snowflake/TempTable.scala (hunk 1 above)
Comment (cryogenian) [reply core/src/main/scala/quasar/destination/snowflake/TempTable.scala:94]: AFAICT, this is sufficient, there is no chance to have two temp tables simultaneously. 
Do you think having UUID postfix is useful?

File: core/src/main/scala/quasar/destination/snowflake/SnowflakeDestination.scala (hunk 2 above)
Comment (cryogenian) [reply core/src/main/scala/quasar/destination/snowflake/SnowflakeDestination.scala:57]: :+1:

**Comment by wemrysi** [comment]
Started looking at this yesterday, will finish up today.

File: core/src/main/scala/quasar/destination/snowflake/Event.scala [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+  type Nothing1[A] = Nothing
+
+  final case class Commit[A](value: A) extends Event[Nothing1, A]
+  final case class Create[F[_]](value: Stream[F, Byte]) extends Event[F, Nothing]
+
+  // 1024 chunks is approximately 8M rows, should be OKayish for backpressure
+  val QueueSize: Int = 1024
+  val ResultQueueSize: Int = 256
+
+  def fromDataEvent[F[_]: Concurrent, A]: Pipe[F, DataEvent[Byte, A], Event[F, A]] = {
```
Comment (wemrysi) [inline core/src/main/scala/quasar/destination/snowflake/Event.scala:42]: Would a different output structure make this easier to use elsewhere? Something like `Out[A](bytes: Stream[F, Byte], commit: F[A])`?

**Review by wemrysi — COMMENTED** [review]
@cryogenian I still have a bit more to look through, but here are some initial comments. Would you mind adding a brief description of what the PR is intending to accomplish? I just want to make sure I understand the goal given the amount of indirection involved.

File: core/src/main/scala/quasar/destination/snowflake/Event.scala [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+sealed trait Event[+F[_], +A]
+
+object Event {
+  type Nothing1[A] = Nothing
+
+  final case class Commit[A](value: A) extends Event[Nothing1, A]
+  final case class Create[F[_]](value: Stream[F, Byte]) extends Event[F, Nothing]
+
+  // 1024 chunks is approximately 8M rows, should be OKayish for backpressure
+  val QueueSize: Int = 1024
```
[the remark this answers is not in the document]
Comment (cryogenian) [reply core/src/main/scala/quasar/destination/snowflake/Event.scala:39]: I don't think that we can compare producing and consuming speeds empirically 🤔 
I'm happy to make smallish though, what do you think it should be? Like 8K? 128K?

File: core/src/main/scala/quasar/destination/snowflake/Event.scala (hunk 3 above)
Comment (cryogenian) [reply core/src/main/scala/quasar/destination/snowflake/Event.scala:42]: I considered using something like `Out[A}(bytes: Stream[F, Byte], offset: A)`, found it a bit inconvenient, `F[A]` might be much better. Thanks :+1:

File: core/src/main/scala/quasar/destination/snowflake/StageFile.scala [hunk 5]
```diff
@@ -19,48 +19,90 @@ package quasar.destination.snowflake
...
-      _ <- blocker.delay[ConnectionIO, Unit](connection.uploadStream("@~", "/", inputStream, name, Compressed))
-      _ <- debug(s"Finished staging to file: @~/$name")
-    } yield new StageFile {
-      def fragment = Fragment.const0(name)
-    }
+      val acquire: F[StageFile] = for {
+        unique <- Sync[F].delay(UUID.randomUUID.toString)
+        uniqueName = s"precog-$unique"
+        _ <- debug(s"Starting staging to file: @~/$uniqueName")
+        _ <- blocker.delay[F, Unit](connection.uploadStream("@~", "/", inputStream, uniqueName, Compressed))
```
[the remark this answers is not in the document]
Comment (cryogenian) [reply core/src/main/scala/quasar/destination/snowflake/StageFile.scala:58]: :+1:

**Comment by cryogenian** [comment]
This pr fixes performance problem with current master. It appears that staging smallish files created from chunks is _much_ slower than using the whole input stream as a file and staging it (for 5M rows I got difference around x60). So, the idea is to go through `Stream[F, DataEvent[Byte, A]]` and create `Stream[F, Event[F, A]]` where `Event` is either `Create[F[_]](value: Stream[F, Byte])` or `Commit[A](value: A)`. 
This is done as follows
+ we create a couple of refs handling a queue of resulting `Event` and a stream of output data in `Create` (wrapped in `Option`)
+ when `DataEvent.Create`  appears in data event stream and queues are `Some` we enqueue it into `Create` queue
+ when `DataEvent.Create` appears in data event stream and queus are `None` we initialize inner queue, then put it to ref and enqueue dequeued inner queue wrapped with `Event.Create` to out queue
+ when `DataEvent.Delete` appears in data event stream we ignore it
+ when `DataEvent.Commit` appears in data event stream we put `Event.Commit` to out stream, and set inner queue to `None`

But this wouldn't work if we just `.through(pipe)`, and so, we grab out queue, dequeue it and `concurrently` put data from data event stream into the queue. This way we can have inner streams being opened until `Commit` appears in input stream. 

The input files for staging is large again and things work with the same performance as they used to.

**Comment by cryogenian** [comment]
Tricky parts and things I was stuck for some time

+ I was going to have lots of `Event.Create` per `DataEvent.Create` continuous region, where the data either initially aggregated into `Event.Create` and then it's split into multiple `Event.Create`. 
+ `Event.Create` flatmapped into `Stream.resource(StageFile.apply(...` this is kinda OK, but I couldn't manage `TempTable` to outlive things, I bet that I missed something simple and obvious (like removing temp table after `persist` or something like that). The idea is to upload data and ingest them in larger chunks ~4-64Gb, with immediate removal of staging files right after consumption.

**Comment by cryogenian** [comment]
Ongoing things that would be added either in this pr or the next one. 

Here `Retry` is removed because it doesn't have much sense to retry streaming data which is already unavailable. Going to restore it, with chunking ~4-8Gb, anyways we handle this amount of data in ram during sliding. The idea is to put the bytes into `Resource[F, Array[Byte]]]` and flatmap them into `StageFile`, then ingest that file into temp table and repeat until we have `Commit` message.

**Comment by wemrysi** [comment]
@cryogenian Assuming it works (I haven't compiled), something like the following might be a bit more direct way to construct the regions

```scala
  final case class Region[F[_], A](data: Stream[F, Byte], commit: F[A])

  def regionPipe(in: Stream[F, DataEvent[F, A]]): Stream[F, Region[F]] =
    Stream.force(Queue.bounded[F, Option[Chunk[Byte]]](128) { q =>
      val region =
        Stream.eval(Deferred[F, A]) flatMap { commit =>
          q.dequeue.unNoneTerminate.peek flatMap {
            case None => Pull.done
            case Some((_, s)) =>
              val bytes = s.flatMap {
                case Create(c) => Stream.chunk(c)
                case Delete(_) => Stream.empty
                case Commit(a) => Stream.eval(commit.complete(a)).drain
              }

              Pull.output1(Region(bytes.interruptWhen(commit.get.void.attempt), commit.get))
          }.stream
        }

      region.repeat.concurrently(in.chunks.noneTerminate.through(q.enqueue))
    })
```

File: core/src/main/scala/quasar/destination/snowflake/TempTable.scala (hunk 1 above)
Comment (rintcius) [reply core/src/main/scala/quasar/destination/snowflake/TempTable.scala:94]: Nah, should be ok without indeed.

File: core/src/main/scala/quasar/destination/snowflake/TempTable.scala [hunk 6]
```diff
@@ -0,0 +1,255 @@
...
+          tmpFragment ++
+          createColumnFragment
+       }
+
+        val dropFragment =
+          fr"DROP TABLE IF EXISTS" ++ tmpFragment
+
+        val acquire = for {
+          _ <- execFragment(createFragment)
+          stageResults <- stages.evalMap(x => execFragment(ingestFragment(x))).compile.toList
```
Comment (wemrysi) [inline core/src/main/scala/quasar/destination/snowflake/TempTable.scala:135]: Similar case here with `acquire`, if we can move this out of the critical region, it'd be preferable from a cancelation perspective.

File: core/src/main/scala/quasar/destination/snowflake/StageFile.scala [hunk 7]
```diff
@@ -19,48 +19,90 @@ package quasar.destination.snowflake
...
+  }
 
-    Resource.make(acquire)(release)
+  def eventPipe[F[_]: ConcurrentEffect: ContextShift, A](
+      connection: SnowflakeConnection,
+      offsets: Queue[F, Option[A]],
+      blocker: Blocker,
+      xa: Transactor[F],
+      logger: Logger)
+      : Pipe[F, Event[F, A], Stream[F, StageFile]] = {
```
Comment (wemrysi) [inline core/src/main/scala/quasar/destination/snowflake/StageFile.scala:80]: Can we avoid the accumulation in favor of a `Pipe[F, Event[F, A], StageFile]`? I may be missing why it is necessary.

File: core/src/main/scala/quasar/destination/snowflake/Event.scala (hunk 4 above)
[the remark this answers is not in the document]
Comment (cryogenian) [reply core/src/main/scala/quasar/destination/snowflake/Event.scala:39]: Ok :+1:

File: core/src/main/scala/quasar/destination/snowflake/StageFile.scala (hunk 7 above)
Comment (cryogenian) [reply core/src/main/scala/quasar/destination/snowflake/StageFile.scala:80]: :+1:

File: core/src/main/scala/quasar/destination/snowflake/TempTable.scala (hunk 6 above)
Comment (cryogenian) [reply core/src/main/scala/quasar/destination/snowflake/TempTable.scala:135]: :+1:

**Comment by cryogenian** [comment]
@wemrysi 

```scala
final case class Region[F[_], A](data: Stream[F, Byte], commit: F[A])
```
I see, you want to emphasize that we always have only one `Commit` message :+1: This simplifies things a lot, thanks!

**Comment by wemrysi** [comment]
@cryogenian Yeah, exactly, it seems like making that explicit might simplify things a bit.

**Comment by cryogenian** [comment]
@rintcius @wemrysi This is ready for another look. 

+ I've switched to something like `Region`, but with multiple possible offsets, in the end managing `Deferred[F, Option[A]]` is almost the same as `InspectableQueue[F, Option[A]]`. 
+ Moved `blocker` to destination constructor arguments

I tried different inner queue sizes, and doesn't seem that it affects performance much, i.e. it should affect if reading from source is much faster than sending via http, but even for local datasource there is no difference. 

Tried multiple `Pull` based approach but nothing worked out as expected, it's either simple and beatiful but doesn't work or ugly stateful and work (I suppose this is what happens when things become concurrent all the time).

**Pull request merged**